camayoc/api.py

```diff
@@ -0,0 +1,224 @@
+# coding=utf-8
+"""Client for working with QCS's API.
+
+This module provides a flexible API client for talking with the quipucords
+server, allowing the user to customize how return codes are handled depending
+on the context.
+
+"""
+
+import requests
+
+from urllib.parse import urljoin
+
+from camayoc import config
+from camayoc import exceptions
+
+
+def echo_handler(server_config, response):  # pylint:disable=unused-argument
+    """Immediately return ``response``."""
+    return response
+
+
+def code_handler(server_config, response):  # pylint:disable=unused-argument
+    """Check the response status code, and return the response.
+
+    :raises: ``requests.exceptions.HTTPError`` if the response status code is
+        in the 4XX or 5XX range.
+    """
+    response.raise_for_status()
+    return response
+
+
+def json_handler(server_config, response):
+    """Like ``code_handler``, but also return a JSON-decoded response body.
+
+    Do what :func:`camayoc.api.code_handler` does. In addition, decode the
+    response body as JSON and return the result.
+    """
+    response.raise_for_status()
+    return response.json()
+
+
+class Client(object):
+    """A client for interacting with the quipucords API.
+
+    This class is a wrapper around the ``requests.api`` module provided by
+    `Requests`_. Each of the functions from that module are exposed as methods
+    here, and each of the arguments accepted by Requests' functions are also
+    accepted by these methods. The difference between this class and the
+    `Requests`_ functions lies in its configurable request and response
+    handling mechanisms.
+
+    All requests made via this client use the base URL of the qcs server
+    provided in your ``$XDG_CONFIG_HOME/camayoc/config.yaml``.
+
+    You can override this base url by assigning a new value to the url
+    field.
+
+    Example::
+        >>> from camayoc import api
+        >>> client = api.Client()
+        >>> # I can now make requests to the QCS server
+        >>> # using relative paths, because the base url is
+        >>> # was set using my config file.
+        >>>
+        >>> client.get('/credentials/hosts/')
+        >>>
+        >>> # now if I want to do something else,
+        >>> # I can change the base url
+        >>> client.url = 'www.whatever.com'
```

Comment:
I need to update all example urls to include `http://` or `https://` prefixes because otherwise they won't work.